### ISSUE TITLE
Vulkan driver: Application attempts to transition a non-presentable image to PRESENT_SRC layout at vkCmdEndRenderPass() call time

### ISSUE THREAD

**Issue by kbiElude** [baldurk/renderdoc#209]
Let's consider a renderpass which uses a single color attachment, and a framebuffer, whose the only color attachment is one of the swapchain images. The renderpass defines an initialLayout value of COLOR_ATTACHMENT_OPTIMAL and a PRESENT_SRC_KHR finalLayout for that attachment. The two objects are later bound together to the GFX pipeline at vkCmdBeginRenderPass() call time, and a full-screen quad is drawn.

Now, if I capture a frame from that app, this works fine. However, once I close the test app and let RenderDoc start loading that frame, I'm running into an issue, where RenderDoc uses a non-presentable image, instead of the original swap-chain image, when configuring the framebuffer. This renders the implicit transition invalid, leading to bad stuff.

**Comment by baldurk** [comment]
Ah right, yes the renderpass createinfo structures need the same fake-presentable image layout patching as transitions etc. Easy enough :smile:.

You mention a 'implicit transition' though - is there actually an implicit transition done by starting/ending a render pass? I thought the createinfo was just a promise from the application that it would ensure the image is found in X state and left in Y state so the driver can reason about it without needing to wait until submit time to see what the actual layout ends up being.

**Comment by kbiElude** [comment]
As per spec:

```
finalLayout is the layout the attachment image subresource will be transitioned to when a render pass instance ends. During a render pass instance, an attachment can use a different layout in each subpass, if desired.
```

In other words, the driver will keep track of what image layout  each subresource was transitioned to in the renderpass, and move the subresource to _finalLayout_, if it has not been moved to that layout already.

There's also a bit more elaborate description of the process in the "Render pass creation" chapter of the spec. It does take time to get one's mind around though.

**Comment by baldurk** [comment]
Interesting, I had completely missed that concept with the SUBPASS_EXTERNAL dependencies helping define implicit transitions. That means RenderDoc's internal image layout tracking is going to be off.

That needs to be fixed too then :).

**Comment by baldurk** [comment]
I opened up a new bug to track the lacking renderdoc tracking of image layouts since I don't think I'll fix it pre-GDC, but the simpler thing of just patching the initial/final layout so it doesn't contain PRESENT_SRC is done now.

**Comment by baldurk** [comment]
Closing this issue as the bug in the original issue should be fixed in [v0.30](https://github.com/baldurk/renderdoc/releases/tag/v0.30). Feel free to re-open if it's not as you expect!
